rust/src/builders/tx_builder.rs

```diff
@@ -485,19 +485,23 @@ impl TransactionBuilder {
                 for i in associated.iter_mut() {
                     let random_index = rng.gen_range(0..relevant_indices.len());
                     let j: &mut usize = relevant_indices.get_mut(random_index).unwrap();
-                    let input = &available_inputs[*i];
-                    let new_input = &available_inputs[*j];
-                    let cur = from_bignum(&input.output.amount.coin);
-                    let new = from_bignum(&new_input.output.amount.coin);
-                    let min = from_bignum(&output.amount.coin);
-                    let ideal = 2 * min;
-                    let max = 3 * min;
-                    let move_closer = (ideal as i128 - new as i128).abs() < (ideal as i128 - cur as i128).abs();
-                    let not_exceed_max = new < max;
-                    if move_closer && not_exceed_max {
-                        std::mem::swap(i, j);
+                    let should_improve = {
+                        let input = &available_inputs[*i];
+                        let new_input = &available_inputs[*j];
+                        let cur = from_bignum(&input.output.amount.coin);
+                        let new = from_bignum(&new_input.output.amount.coin);
+                        let min = from_bignum(&output.amount.coin);
+                        let ideal = 2 * min;
+                        let max = 3 * min;
+                        let move_closer = (ideal as i128 - new as i128).abs() < (ideal as i128 - cur as i128).abs();
+                        let not_exceed_max = new < max;
+
+                        move_closer && not_exceed_max
+                    };
+                    if should_improve {
                         available_indices.insert(*i);
                         available_indices.remove(j);
+                        std::mem::swap(i, j);
```

Comment:
This should fix it. New input should be removed from available UTxO set whereas the old one should be put back. The order didn't make sense. @SebastienGllmt

Reply:
@rooooooooob

Reply:
Good catch @siegfried

Reply:
After #53 is merged, this is all set. Or merge this directly since it has already rebased on that one.

Reply:
I think that I started writing it another way then later changed to indices halfway through since it made something easier. I forgot what though.

Reply:
Ok, thanks. I think this is ready to merge. I can create another PR to refactor if it is needed.